### DISCUSSION TITLE
accessMode/volume Mode

### DISCUSSION THREAD

**Discussion by lmcdasm** [topolvm/topolvm#663] [Q&A] [answered]
Hello there.

I was curious about how to set the accessMode and volumeMode in the storage class (or perhaps i should not and do it explicitly in the POD or elsewhere?) and how to know what is supportted - i saw in the csi/csi.go there is code for the different Volume and Access modes, however i wasnt sure if i can read that the underlying device-class and fs-type is what determines what value to put?

Some pointers are helpful!
cheers,
D

**Comment by llamerada-jp** [comment, marked as answer]
Hi lmcdasm,

TopoLMV supports below modes in PVC. [Filesystem format (ext4, xfs...)](https://github.com/topolvm/topolvm/blob/main/docs/user-manual.md#storageclass) can be specify in SC of TopoLVM.
- accessMode: ReadWriteOnce / ReadWriteOncePod
- volumeMode: Filesystem / Block

About accessMode, you may able to specify ReadOnlyMany or ReadWriteMany. TopoLVM does not explicitly reject these values. But PVs made by TopoLVM are bonded to the node having actual lv. So practically ReadOnlyMany and ReadWriteMany are nonsense.
If you are interested in the CSI driver implementation, please refer to the following document.
- [Container Storage Interface (CSI) - CreateVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume)
  CreateVolumeRequest.volume_capabilities.access_mode is specified based on accessMode of PVC. You should return InvalidArgument error when the specified mode is unsupported by the CSI driver.
- [KEP-2485: ReadWriteOncePod PersistentVolume AccessMode](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode)
- [Kubernetes CSI Developer Documentation - 2.6.3. Raw Block Volume](https://kubernetes-csi.github.io/docs/raw-block.html)

**Comment by lmcdasm** [comment]
Hello @llamerada-jp 

Thanks so much! Thats exactly what i was on the hunt for and wasnt able to find it so clearly laid out!

I was looking to get to the underlying implementation yes, so you just helped me with both, let me digest!

Amazing, thanks again!
Daniel
